### PR TITLE
feat(memory): add SQLite storage foundation

### PR DESCRIPTION
**Pull request by YayoiNanoka**
## Summary

Implements the SQLite storage foundation for the long-term-memory design tracked in #1615.

- define atomic memory-item contracts for lifecycle, temporal semantics, scope, provenance, and normalized retrieval keys
- add a dedicated `<stateRoot>/memory.sqlite` store with transactional migrations, optimistic concurrency, idempotent batch mutations, exact/prefix key search, and fail-closed corruption checks
- require authenticated Interactive or Headless Storage Root leases to open the store
- make Runtime Host startup and shutdown own the database lifecycle
- exclude the database and its sidecars from Session Bundles and extend the Runtime Host and Headless dependency boundaries

## Scope

Refs #1615. This is PR ① from the [accepted design resolution](https://github.com/maka-agent/maka-agent/issues/1615#issuecomment-5139323816).

This PR establishes the durable `memory_item` fact and evidence layer only. It intentionally does not add Runtime extraction, group/digest projections, `memory_search`, prompt injection, recall, or benchmark wiring. No production path writes or recalls actual memories yet.

## Verification

- `@maka/core` full suite: 1,254 passed
- `@maka/storage` full suite: passed
- long-term-memory and Runtime Host focused tests: 30/30 passed
- Runtime Host dependency boundary: 5/5 passed
- Headless storage dependency boundary: 3/3 passed
- `biome lint`, `biome format`, and `git diff --check`: passed

## Review focus

- schema, migration, transaction, and corruption boundaries in the new SQLite store
- Storage Root authority checks and Runtime Host open/close ownership
- Session Bundle exclusion and the absence of production memory read/write wiring

<details>
<summary>中文说明</summary>

## 概要

本 PR 实现 #1615 长期记忆设计中的 SQLite 存储基础。

- 定义原子化 `memory_item` 契约，包括生命周期、时间语义、作用域、来源和规范化检索键
- 增加独立的 `<stateRoot>/memory.sqlite`，支持事务化迁移、乐观并发控制、幂等批量写入、关键词精确/前缀检索，以及针对损坏数据的 fail-closed 校验
- 只有持有经过认证的 Interactive 或 Headless Storage Root lease 才能打开该数据库
- 由 Runtime Host 负责数据库的启动和关闭生命周期
- 从 Session Bundle 中排除数据库及其 sidecar 文件，并扩展 Runtime Host 与 Headless 的依赖边界

## 范围

关联 #1615。这是[已确认设计方案](https://github.com/maka-agent/maka-agent/issues/1615#issuecomment-5139323816)中的 PR ①。

本 PR 只建立持久化的 `memory_item` 事实与证据层，暂不包含 Runtime 记忆提取、group/digest 投影、`memory_search`、Prompt 注入、召回和 benchmark 接线。目前没有生产路径会实际写入或召回记忆。

## 验证

- `@maka/core` 全量测试：1,254 passed
- `@maka/storage` 全量测试：passed
- 长期记忆和 Runtime Host 聚焦测试：30/30 passed
- Runtime Host 依赖边界：5/5 passed
- Headless Storage 依赖边界：3/3 passed
- `biome lint`、`biome format` 和 `git diff --check`：passed

## Review 重点

- 新 SQLite Store 的 schema、迁移、事务和损坏数据处理边界
- Storage Root 权限校验，以及 Runtime Host 对数据库打开/关闭生命周期的所有权
- Session Bundle 排除规则，以及本 PR 没有接入生产记忆读写

</details>

### CODE REVIEW

**Comment by Astro-Han** [comment]
I read the full diff against the existing store and facade conventions, and ran a few empirical checks (lone surrogates through node:sqlite, `user_version` transactionality, busy_timeout). No P0 or P1 blockers. Two P2s are worth deciding on before the next PR, the schema validation will break on the first ALTER migration, and there are test gaps. Roughly 15-20% of the diff belongs to later PRs.

## P2: content normalizer accepts lone surrogates

`normalizeLongTermMemoryContent` (packages/core/src/long-term-memory.ts:214) strips C0/C1 and zero-width characters but not `\p{Cs}`. `normalizeVisibleText` (packages/storage/src/sqlite-long-term-memory-store.ts:834) rejects them, so the two normalizers disagree inside the same store boundary.

Verified: writing `'lone\uD800surrogate'` through node:sqlite round-trips as U+FFFD. The read-back hash check passes because Node's utf8 encoder applies the same lossy replacement, so the store accepts the write and silently changes the content. The observable break is `semanticItemMatchesWrite` (store:921): a second identical write compares the raw surrogate against the stored U+FFFD string, finds no match, and creates a duplicate item.

No production path writes today and LLM output with a lone surrogate is rare, so this is not a blocker. Fix it before the first real writer lands: reject `\p{Cs}` in the content normalizer (aligns with `normalizeVisibleText`) and add one regression test. The legacy `normalizeMemoryContent` (packages/core/src/memory.ts:322) has the same hole; same fix if you touch it.

## P2: semantic dedup in the store layer drops new evidence

`#createItem` (store:317) returns `existing` when `#findActiveSemanticDuplicate` hits, and the new mutation's keys and sources are discarded. The test at sqlite-long-term-memory-store.test.ts:250 pins this ("without merging sources"). Issue #1615 treats RuntimeEvent as the evidence source, so a fact re-extracted from a new session loses its new provenance unless the extraction layer manually merges before update.

The bigger question is placement: "same content + kind + statement + temporal + scope = same fact" is an extraction-layer decision, not a storage invariant. Hardcoding it here fixes the strategy, including the evidence-dropping behavior, before any caller exists. Either move dedup to the extraction PR (the store keeps a partial unique index on scope + content_hash + lifecycle), or make `existing` signal that evidence was not merged.

## P2: schema validation is pinned to version 1

`REQUIRED_SCHEMA_DEFINITIONS` is extracted from `MIGRATIONS.get(1)!` and `validateSqliteLongTermMemorySchema` runs on every open. The first migration that does `ALTER TABLE ... ADD COLUMN` changes `sqlite_master.sql`, and `assertSchemaObject` requires normalized SQL-text equality plus exact column order (schema:129-172, 313-339). That open throws `invalid definition for table memory_items` with no hint that it is a versioning problem. Derive expectations from migration history up to the observed user_version before PR ②.

Separately, full-DDL validation on every open is a new precedent; the other sqlite stores (runtime, session metadata, usage) only check user_version. If the validation earns its place, extract it as a shared helper. Otherwise trim to the same level.

## Test gaps

The store suite is strong on rollback boundaries, corruption, first-open migration concurrency, and CAS. Missing:

- `invalid_lifecycle_transition` and `item_not_found` appear nowhere in the suite. Archive-of-archived, restore-of-active, and update/archive/restore of a missing item are all untested.
- `duplicate_active` on updating an active item, the daily path, is untested; only archived-update and restore collisions are.
- Multi-term search semantics: AND/ranking, term dedup, default limit, and the 1..100 clamp are untested. Every search test uses a single term.
- No store-level write with point/interval/open_ended bounds. `open_ended` with an end is untested at any level. `observedAt > committedAt` rejection is untested.
- The input-validation layer (empty or oversized mutations, conflicting source provenance, control chars in keys, non-NFC identifiers) has no tests.
- No durability test. sqlite-runtime-crash.test.ts is the repo precedent; nothing here verifies a committed receipt survives checkpoint + reopen, or a kill between COMMIT and return.
- No truly concurrent applyMutations from two connections; the existing two-connection test is sequential.
- The core contract test is thin: temporal boundaries now == start / now == end, the 2000/2001 code-point edges, and lone-surrogate content are unpinned.

Two weak tests: "bounds corrupt Item child reads before checking cardinality" (test:592) passes even with the LIMIT removed, so it does not guard what it claims; the EXPLAIN QUERY PLAN regex pins SQLite plan text for a query that already carries INDEXED BY, so it guards an annotation more than behavior.

## Scope: what belongs in this PR

A few pieces look like later-PR material:

- The composition wiring (execution-composition.ts:73) creates memory.sqlite plus sidecars on every startup with zero readers or writers, and forces the bundle-policy exclusions and two lifecycle tests. If no production writer is near, move this plus the session-bundle exclusions to that PR. If you keep it, state the benefit (e.g. surfacing lease integration early) in the description.
- `deriveMemoryTemporalStatus` has no caller outside its own test. The temporal invariants now live in three places (schema CHECK, `validateTemporalBounds`, this function) across two packages. Give it a consumer or share one validator.
- `memory_items_by_kind` has no query method using it.
- `user_requested` origin is forced to carry at least one source Event, but a user-written preference has no RuntimeEvent. Loosen or document before the extraction PR.
- The facade is the 7th copy of the single-flight pattern, and the only one with the generic-K shape plus a headless arm that the headless dependency test forbids production code from importing. shell-run-authority.ts remains the closest seam; either copy its shape or extract the shared helper and converge the family.

The idempotency receipts and CAS versioning, by contrast, look right for the store layer and should stay.

<details>
<summary>中文版</summary>

我把整个 diff 对照仓库里现有的 store 和 facade 惯例通读了一遍，并做了几个实证验证（孤立代理项经 node:sqlite 的往返、`user_version` 的事务性、busy_timeout 行为）。没有 P0 或 P1 的阻塞项。有三个问题值得在下一个 PR 之前决策：两个 P2、以及一个会在第一次 ALTER 迁移时崩掉的 schema 校验。此外还有一批测试缺口。粗略估计整个 diff 里有 15-20% 属于后续 PR 的范围。

**P2：内容规范化器接受孤立代理项**

`normalizeLongTermMemoryContent`（packages/core/src/long-term-memory.ts:214）清理了 C0/C1 控制字符和零宽字符，但没有拒绝 `\p{Cs}`（孤立代理项）。而 `normalizeVisibleText`（packages/storage/src/sqlite-long-term-memory-store.ts:834）拒绝它们，同一个 store 内部两个规范化器口径不一致。

实证结果：写入 `'lone\uD800surrogate'` 经 node:sqlite 往返后变成 U+FFFD。读回时的哈希校验仍然通过，因为 Node 的 utf8 编码器做了同样的有损替换，于是 store 接受这次写入并悄悄改了内容。真正可观察的破坏点在 `semanticItemMatchesWrite`（store:921）：第二次写入相同内容时，原始代理项字符串和库里存着的 U+FFFD 字符串比较不相等，判定为不重复，就会创建重复 item。

当前没有任何生产路径会写入，LLM 输出带孤立代理项也罕见，所以这不是阻塞项。但应该在第一个真实写入方落地前修掉：在内容规范化器里拒绝 `\p{Cs}`（与 `normalizeVisibleText` 对齐），再加一条回归测试。legacy 的 `normalizeMemoryContent`（packages/core/src/memory.ts:322）有同样的洞，顺手一起修即可。

**P2：store 层的语义去重会静默丢弃新证据**

`#createItem`（store:317）在 `#findActiveSemanticDuplicate` 命中时返回 `existing`，新 mutation 携带的 keys 和 sources 被丢弃。sqlite-long-term-memory-store.test.ts:250 的测试明确钉住了这个行为（测试名就叫 "without merging sources"）。#1615 把 RuntimeEvent 当作证据来源，那么同一事实从新 session 再次提取时，除非 extraction 层先手动合并再 update，否则新来源就丢了。

更深一层的问题是分层："内容 + kind + statement + temporal + scope 相同 = 同一条事实"是 extraction 层的决策，不是存储不变量。把它硬编码在这里，等于在没有任何调用方之前就锁死了这个策略，包括丢证据这个行为。要么把去重移到 extraction PR（store 只保留 scope + content_hash + lifecycle 的 partial unique index），要么让 `existing` 结果明确带上"证据未合并"的信号。

**P2：schema 校验被钉死在版本 1**

`REQUIRED_SCHEMA_DEFINITIONS` 从 `MIGRATIONS.get(1)!` 提取，`validateSqliteLongTermMemorySchema` 每次打开都会跑。第一次做 `ALTER TABLE ... ADD COLUMN` 的迁移会让 `sqlite_master.sql` 变化，而 `assertSchemaObject` 要求归一化后的 SQL 文本完全相等、列顺序完全一致（schema:129-172, 313-339）。到那时每次打开都会抛 `invalid definition for table memory_items`，且没有任何线索提示这是版本问题。建议在 PR ② 之前改成按观察到的 user_version 从迁移历史推导期望值。

另外，每次打开都做全量 DDL 校验是仓库里的新先例，其他 sqlite store（runtime、session metadata、usage）都只查 user_version。如果这套校验确实值得保留，就抽成共享 helper；否则砍到和其他 store 同级。

**测试缺口**

store 测试套件在回滚边界、坏数据、首次打开迁移并发、CAS 上都很扎实。缺的是：

- `invalid_lifecycle_transition` 和 `item_not_found` 在整个套件里一次都没出现。已归档再归档、active 项 restore、对不存在的 item 做 update/archive/restore，全部没测。
- 更新 active item 撞 `duplicate_active`（日常路径）没测，只测了归档项更新和 restore 碰撞。
- 多词搜索语义没测：AND/排序、term 去重、默认 limit、1..100 钳制。所有搜索测试都只用单个词。
- store 层没有带 point/interval/open_ended 边界的写入测试。`open_ended` 带 end 在任何层级都没测。`observedAt > committedAt` 的拒绝也没测。
- 输入校验层（空/超量 mutations、来源 provenance 冲突、key 里的控制字符、非 NFC 标识符）完全没有测试。
- 没有持久性测试。仓库有 sqlite-runtime-crash.test.ts 先例；这里没有任何东西验证已提交的 receipt 在 checkpoint + 重开之后、或在 COMMIT 与返回之间被杀进程之后仍然存活。
- 没有两个连接真正并发的 applyMutations；现有的双连接测试是顺序执行的。
- core 契约测试偏薄：时间边界 now == start / now == end、2000/2001 码点边界、孤立代理项内容都没有钉住。

另外两个弱测试："bounds corrupt Item child reads before checking cardinality"（test:592）即使删掉 LIMIT 也能通过，它并没有守护它声称的东西；EXPLAIN QUERY PLAN 的正则钉的是 SQLite 计划文本，而查询本身就带了 INDEXED BY，所以它守护的是实现注解而非行为。

**范围：哪些属于本 PR**

有几块看起来属于后续 PR：

- composition 接线（execution-composition.ts:73）在每次启动时创建 memory.sqlite 加 sidecar 文件，全程零读写，还连带逼出了 bundle-policy 排除项和两个生命周期测试。如果近期没有生产写入方，把这段连同 session-bundle 排除项一起移到那个 PR 才划算。如果坚持保留，请在描述里说明收益（比如提前暴露 lease 集成问题）。
- `deriveMemoryTemporalStatus` 除了自己的测试没有任何调用方。时间不变量现在散落在三处（schema CHECK、`validateTemporalBounds`、这个函数），跨两个包。要么给它一个消费者，要么抽一个共享校验器。
- `memory_items_by_kind` 索引没有任何查询方法使用它。
- `user_requested` origin 被强制要求至少携带一个 source Event，但用户手写的偏好没有 RuntimeEvent。在 extraction PR 之前放宽或写清楚。
- facade 是单飞模式的第 7 份拷贝，而且是唯一采用泛型 K 形状、并且 headless 依赖测试禁止生产代码导入其 headless arm 的那份。shell-run-authority.ts 仍然是最接近的接缝；要么照抄它的形状，要么抽共享 helper 把整个家族收敛过去。

相比之下，幂等 receipt 和 CAS 版本号放在 store 层是对的，应该保留。

</details>

**Comment by YayoiNanoka** [comment]
Thanks for the detailed review. I’ve pushed the follow-up changes in `b13fc833`.

The main changes are:

- Reject lone Unicode surrogates at the long-term memory content boundary.
- Keep exact semantic duplicate detection in the Store, but return `duplicate_active` with the existing `conflictingItemId`. The Store no longer silently drops or merges new evidence; PR ② will handle read/merge/CAS explicitly.
- Use `PRAGMA user_version` as the migration authority and remove full DDL-text validation on every open.
- Remove the unused `deriveMemoryTemporalStatus`; temporal bounds now use one shared Core validator, with SQLite `CHECK` constraints as the persistence safety net.
- Remove the unused `memory_items_by_kind` index. Query-shaped indexes can be added when corresponding access paths exist.
- Keep `user_requested` provenance tied to a `RuntimeEvent`, since explicit user requests enter the runtime as events.
- Simplify the facade to the Interactive Runtime Host path only. Headless integration remains deferred to the benchmark/extraction work.
- Keep Runtime Host lifecycle wiring in PR ① so opening, migration, lease validation, startup-failure cleanup, and shutdown behavior are exercised before production writers arrive.
- Expand lifecycle, collision, multi-term search, temporal validation, malformed input, durability, crash, and concurrent CAS coverage.

During regression testing, I also found and fixed a real SQLite sidecar race: a WAL/SHM file can become unlinked between opening and validation during concurrent initialization. Regenerable sidecars now tolerate that specific `nlink = 0` state, while the main database still requires exactly one link.

Validation completed successfully:

- Core: 765 tests passed
- Long-term memory Storage: 31 tests passed
- Runtime Host lifecycle: 5 tests passed
- Headless dependency boundary: 3 tests passed

<details>
<summary>中文翻译</summary>

感谢详细的 review。相关修改已经推送到 `b13fc833`。

主要修改如下：

- 在长期记忆内容边界拒绝不完整的 Unicode surrogate。
- Store 继续检测完全相同的活跃记忆，但现在会返回包含现有 `conflictingItemId` 的 `duplicate_active`。Store 不再静默丢弃或合并新证据；PR ② 将显式执行读取、合并和 CAS 更新。
- 使用 `PRAGMA user_version` 作为迁移版本的事实来源，删除每次打开数据库时进行的完整 DDL 文本校验。
- 删除暂时没有调用方的 `deriveMemoryTemporalStatus`。时间边界统一使用 Core 中的共享校验器，SQLite `CHECK` 约束作为持久化层的最后保护。
- 删除没有对应查询路径的 `memory_items_by_kind` 索引。后续根据实际查询方式增加相应索引。
- `user_requested` 仍然要求关联 `RuntimeEvent`，因为用户明确提出的记忆请求本身也会进入 RuntimeEvent。
- facade 目前只保留 Interactive Runtime Host 路径；Headless 接入留到后续 benchmark 或提取相关工作。
- PR ① 继续保留 Runtime Host 生命周期接线，用于提前验证数据库打开、迁移、lease 校验、启动失败清理和正常关闭行为。
- 补充了生命周期、重复冲突、多关键词搜索、时间校验、非法输入、持久性、进程崩溃和并发 CAS 等测试。

回归测试期间还发现并修复了一个真实的 SQLite sidecar 竞态：并发初始化时，WAL/SHM 文件可能在打开后、校验前被 SQLite 解除链接。现在仅允许可重新生成的 sidecar 处于 `nlink = 0` 状态，主数据库文件仍然必须保持唯一链接。

验证结果：

- Core：765 项测试通过
- 长期记忆 Storage：31 项测试通过
- Runtime Host 生命周期：5 项测试通过
- Headless 依赖边界：3 项测试通过

</details>

File: packages/storage/src/sqlite-long-term-memory-schema.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+export function migrateSqliteLongTermMemoryDatabase(
+  db: DatabaseSync,
+  options: SqliteLongTermMemoryMigrationOptions = {},
+): void {
+  const observedVersion = readSqliteLongTermMemorySchemaVersion(db);
+  if (observedVersion > SQLITE_LONG_TERM_MEMORY_SCHEMA_VERSION) {
+    throw newerSchemaError(observedVersion);
+  }
+  if (observedVersion === SQLITE_LONG_TERM_MEMORY_SCHEMA_VERSION) return;
```
Comment (Astro-Han) [inline packages/storage/src/sqlite-long-term-memory-schema.ts]: **P2 (non-blocking):** Could we retain a minimal fail-closed check for a current-version database whose required structure is incomplete?

For example, setting `user_version = 1` on an empty database lets the Store open successfully, and the missing table is only reported on the first operation. Similarly, dropping `memory_item_keys_by_normalized_key` is not detected until `searchByKeys()` executes its `INDEXED BY` query.

Since Runtime Host treats a successful open as readiness, please consider validating the version-specific minimum structure after migration: required tables and columns, plus indexes that queries explicitly depend on. This need not restore exact DDL-text comparison. Two regression tests for a current-version database with a missing table and missing required index would cover the boundary.

File: packages/storage/src/sqlite-long-term-memory-store.ts [hunk 2]
```diff
@@ -0,0 +1,1296 @@
...
+
+  #updateItem(
+    mutation: Extract<NormalizedMutation, { type: 'update' }>,
+    mutationIndex: number,
+    committedAt: number,
+    transientItemIds: ReadonlySet<string>,
+  ): MemoryMutationResult {
+    const current = this.#requireVersion(mutation.itemId, mutation.expectedVersion);
+    const currentRecord = this.#requireItemRecord(mutation.itemId);
+    const duplicate = this.#findActiveFactDuplicate(mutation.item, mutation.itemId);
```
Comment (Astro-Han) [inline packages/storage/src/sqlite-long-term-memory-store.ts:389]: **P3 (non-blocking):** Could the exact no-op check happen before active-duplicate detection?

A legal sequence is: archive Item A, create an identical active Item B, then update archived A with a write that exactly matches A’s existing record. This currently returns `duplicate_active`, even though the operation changes nothing and cannot violate the documented “one active row per fact” invariant.

Allowing the no-op before this check would keep restore collisions protected while avoiding an unnecessary restriction on archived records. If the stronger behavior is intentional, the contract and invariant comment should make that explicit.

**Review by Astro-Han — APPROVED** [review]
The storage foundation is in good shape now. The transaction, migration, CAS, idempotency, crash-recovery, and lifecycle paths are well designed, and the expanded tests exercise the important behavior rather than only implementation details.

Approving with two non-blocking follow-ups: one current-version corruption boundary and one archived no-op semantic edge case.
